File: server/src/retrieval/contextCompression.js
// Real implementation — generalized from ragService.js so any agent
// (KnowledgeAgent, ResearchAgent, ...) can reuse the same dedupe/budget
// logic regardless of which search path produced the candidates.

// Removes near-duplicate chunks (same leading text) that retrieval sometimes
// returns when a document has repeated boilerplate or overlapping chunks.
export function dedupeChunks(chunks) {
  const seen = new Set();
  const out = [];
  for (const chunk of chunks) {
    const key = chunk.content.trim().slice(0, 200);
    if (seen.has(key)) continue;
    seen.add(key);
    out.push(chunk);
  }
  return out;
}

// Keeps adding chunks (already sorted by relevance) until the context
// budget is used up, instead of sending every retrieved chunk to the LLM.
export function compressContext(chunks, maxChars) {
  let used = 0;
  const kept = [];
  for (const chunk of chunks) {
    if (used + chunk.content.length > maxChars) break;
    kept.push(chunk);
    used += chunk.content.length;
  }
  return kept;
}

export function estimateTokens(text) {
  return Math.ceil((text || '').length / 4);
}

File: server/app/retrieval/context_compression.py
def dedupe_chunks(chunks: list[dict]) -> list[dict]:
    """Removes near-duplicate chunks (same leading text) that retrieval
    sometimes returns when a document has repeated boilerplate or overlapping chunks."""
    seen = set()
    out = []
    for chunk in chunks:
        key = chunk["content"].strip()[:200]
        if key in seen:
            continue
        seen.add(key)
        out.append(chunk)
    return out


def compress_context(chunks: list[dict], max_chars: int) -> list[dict]:
    """Keeps adding chunks (already sorted by relevance) until the context
    budget is used up, instead of sending every retrieved chunk to the LLM."""
    used = 0
    kept = []
    for chunk in chunks:
        if used + len(chunk["content"]) > max_chars:
            break
        kept.append(chunk)
        used += len(chunk["content"])
    return kept


def estimate_tokens(text: str) -> int:
    return -(-len(text or "") // 4)  # ceil division
